src/main/java/org/assertj/core/internal/DeepDifference.java

```diff
@@ -307,37 +313,58 @@ private static boolean hasCustomComparator(DualKey dualKey, Map<String, Comparat
                                              TypeComparators comparatorByType) {
     if (dualKey.key1.getClass() == dualKey.key2.getClass()) {
       String fieldName = dualKey.getConcatenatedPath();
-      Comparator<?> fieldComparator = comparatorByPropertyOrField.containsKey(fieldName)
-          ? comparatorByPropertyOrField.get(fieldName) : comparatorByType.get(dualKey.key1.getClass());
+      Comparator<?> fieldComparator = comparatorByPropertyOrField != null && comparatorByPropertyOrField.containsKey(fieldName)
+        ? comparatorByPropertyOrField.get(fieldName) : comparatorByType == null
+        ? null : comparatorByType.get(dualKey.key1.getClass());
       return fieldComparator != null;
     }
     return false;
   }
 
-  private static Deque<DualKey> initStack(Object a, Object b, Set<DualKey> visited) {
+  private static Deque<DualKey> initStack(Object a, Object b, List<String> parentPath,
+                                          Map<String, Comparator<?>> comparatorByPropertyOrField,
+                                          TypeComparators comparatorByType) {
     Deque<DualKey> stack = new LinkedList<>();
-    if (a != null && !isContainerType(a)) {
+    List<String> currentPath = parentPath == null ? new ArrayList<String>() : parentPath;
+    DualKey basicDualKey = new DualKey(currentPath, a, b);
+    if (a != null && b != null && (parentPath == null || !hasCustomComparator(basicDualKey, comparatorByPropertyOrField, comparatorByType))
+      && !isContainerType(a) && !isContainerType(b)) {
       // disregard the equals method and start comparing fields
-      Collection<Field> fieldsOfRootObject = getDeclaredFieldsIncludingInherited(a.getClass());
-      if (!fieldsOfRootObject.isEmpty()) {
-        for (Field field : fieldsOfRootObject) {
-          String fieldName = field.getName();
-          DualKey dk = new DualKey(Arrays.asList(fieldName),
-                                   COMPARISON.getSimpleValue(fieldName, a),
-                                   COMPARISON.getSimpleValue(fieldName, b));
-          if (!visited.contains(dk)) {
+      Collection<Field> aFields = getDeclaredFieldsIncludingInherited(a.getClass());
+      if (!aFields.isEmpty()) {
+        Set<String> aFieldsNames = getFieldsNames(aFields);
+        Set<String> bFieldsNames = getFieldsNames(getDeclaredFieldsIncludingInherited(b.getClass()));
+        if (!bFieldsNames.containsAll(aFieldsNames)) {
```

Comment:
Probably here instead of `contansAll` could have been done check on equals of fileds sets, but it corresponds more to how it is working now.